In [75]:
import pandas as pd

In [80]:
df = pd.read_csv('/content/IRIS.csv')

In [77]:
df.head()

,sepal_length,sepal_width,petal_length,petal_width,species
0,5.1,3.5,1.4,0.2,Iris-setosa
1,4.9,3.0,1.4,0.2,Iris-setosa
2,4.7,3.2,1.3,0.2,Iris-setosa
3,4.6,3.1,1.5,0.2,Iris-setosa
4,5.0,3.6,1.4,0.2,Iris-setosa


In [84]:
df.shape

(150, 5)

In [102]:
null_data = df.isnull()
null_data.sum()

,0
sepal_length,0
sepal_width,0
petal_length,0
petal_width,0
species,0


In [99]:
df['species'].value_counts()

,count
species,
Iris-setosa,50
Iris-versicolor,50
Iris-virginica,50


So, by now we've seen that our dataframe has 150 lines, 5 rows and there is no null data, so no need to use the function df.dropna(), and 3 labels with 50 data each.
The target class for our algorithm will be the column species.

In order to elaborate the more appropriate questions at each node, we may take a look at a few statistics of the dataframe:

In [103]:
for specie in df['species'].unique():
  print('\t\t\t' + specie)
  print(df[df['species'] == specie].describe())
  print('\n')


			Iris-setosa
       sepal_length  sepal_width  petal_length  petal_width
count      50.00000    50.000000     50.000000     50.00000
mean        5.00600     3.418000      1.464000      0.24400
std         0.35249     0.381024      0.173511      0.10721
min         4.30000     2.300000      1.000000      0.10000
25%         4.80000     3.125000      1.400000      0.20000
50%         5.00000     3.400000      1.500000      0.20000
75%         5.20000     3.675000      1.575000      0.30000
max         5.80000     4.400000      1.900000      0.60000


			Iris-versicolor
       sepal_length  sepal_width  petal_length  petal_width
count     50.000000    50.000000     50.000000    50.000000
mean       5.936000     2.770000      4.260000     1.326000
std        0.516171     0.313798      0.469911     0.197753
min        4.900000     2.000000      3.000000     1.000000
25%        5.600000     2.525000      4.000000     1.200000
50%        5.900000     2.800000      4.350000     1.300000
75% 

In order the biggests sepal lengths are:
- Setosa, versicolor and virginica: since the minimum of both versicolor and virgnica lenghts are 4.9, a good question could be "Is the sepal length less than 4.9? I yes it is setosa, if not than it is not setosa, go to next node"

- Sepal width:
"Is it bigger than 3.8? then setosa."

- Petal length:
"Is it less than 3? then setosa."

So in this first approach the setosa seems pretty easy to unmix from the other labels, but versicolor and virginica have a few similarities.   

Range of values
- sepal_length: 4.3 to 7.9
- sepal_width: 2.0 to 4.4
- petal_length:1.0 to 6.9
- petal_width: 0.1 to 2.5


## CART algorithm
We're going to use a Classification and Regression Tress algorithm. It basically works as adding a node to the tree, putting all the training data through this node, asking a boolean question, that means, a kind of yes or no question, then we split the data into subsets, that will be the inputs for the following nodes. Our goal is to find the best questions at each node, the questions that best unmix the data. We can do that mesuring how much a question really does that well.


## The Gini Impurity
It measures the probability to incorrectly label an data, randomly, from the dataframe. If the data are all from the same label, the impurity will be 0.
Its formula is 1 - Sum of probabilities of all labels squared.


In [ ]:
def gini(df):
  impurity = 1
  prob_sum = 0
  data_lableds = df['species'].value_counts().to_list()
  probabilities = []
  for value in data_lableds:
    probabilities.append(value/float(df.shape[0]))
  for probability in probabilities:
    probability = probability**2
    prob_sum += probability
  impurity -= prob_sum
  return impurity

In [101]:
df_setosa = df.head(15) # as the first objects of the dataset are all from the same label, the gini should be zero for this frame
print(gini(df_setosa))

0.0


## Information gain
It is basically a number that tells how much a node helps to unmix the labels at a node. It goes as follows:
- First we calculate the impurity of our dataset
- Then for each node, we calculate the impurity of the child nodes, and take a weighted average of that result;
- Finally, we subtract this with the initial impurity;     

That is how we get the information gain.
We will use it to select the best questions at each point of the tree.

The formula is IG = Gini_parent​ − (left_w​eight × Gini_left​ + right_weight ​× Gin_right​)



In [105]:
def information_gain(left_node, right_node, current_impurity):
  weight_left = float(len(left_node)) / (float(len(left_node)) + float(len(right_node)))
  weight_right = 1 - weight_left
  return current_impurity - (weight_left*gini(left_node) + weight_right*gini(right_node))

Range of values

    sepal_length: 4.3 to 7.9
    sepal_width: 2.0 to 4.4
    petal_length:1.0 to 6.9
    petal_width: 0.1 to 2.5


In [ ]:
class Question:
  def __init__(self, column, value):
    self.column = column
    self.value = value

def partition(df, question):
  true_rows = df[df[question.column]] >= question.value
  false_rows = df[df[question.column]] < question.value
  return true_rows, false_rows


Here are the steps to put all together and make the decision tree:
- First, we calculate the initial impurity;
- Then we iterate over each column and over each value between the range for that column;
- With that, we are going to have the best questions and the best root for our tree;
- We use the following best questions to make the entire tree


In [ ]:
def find_best_root(df):
  initial_impurity = gini(df)
  for column in df:
    if column == 'sepal_length':
      values =
    if column == 'sepal_width':

    if column == 'petal_length':

    if column == 'petal_width':


    sepal_length: 4.3 to 7.9
    sepal_width: 2.0 to 4.4
    petal_length:1.0 to 6.9
    petal_width: 0.1 to 2.5


In [113]:

sepal_len = list(range(43, 79))
print(sepal_len)

[43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68, 69, 70, 71, 72, 73, 74, 75, 76, 77, 78]


In [122]:
df['petal_length'] = (df['petal_length'] * 10)
df['sepal_length'] = df['sepal_length'] * 10
df


,sepal_length,sepal_width,petal_length,petal_width,species
0,5100.0,3.5,1400000.0,0.2,Iris-setosa
1,4900.0,3.0,1400000.0,0.2,Iris-setosa
2,4700.0,3.2,1300000.0,0.2,Iris-setosa
3,4600.0,3.1,1500000.0,0.2,Iris-setosa
4,5000.0,3.6,1400000.0,0.2,Iris-setosa
...,...,...,...,...,...
145,6700.0,3.0,5200000.0,2.3,Iris-virginica
146,6300.0,2.5,5000000.0,1.9,Iris-virginica
147,6500.0,3.0,5200000.0,2.0,Iris-virginica
148,6200.0,3.4,5400000.0,2.3,Iris-virginica
